### PR TITLE
Expose internal types as public access

### PR DESCRIPTION
**Pull request by zao111222333**
## Summary

Thanks for providing such a powerful project! 
I try to use `ruff_python_parser` as tokenizer for my toy python REPL project, https://github.com/zao111222333/pyapp.
Everything is so great besides some necessary types/APIs is internal rightnow. And I request expose them. Thanks again!

btw, here is a small example for the python REPL, `ruff_python_parser` is used to highlight & incomplete detect.

## Test Plan

I believe those changes will not change the functionality, so no test plan :)

### CODE REVIEW

File: crates/ruff_python_parser/src/lib.rs [hunk 1]
```diff
@@ -67,12 +67,12 @@
...
 use std::ops::Deref;
 
-pub use crate::error::{FStringErrorType, ParseError, ParseErrorType};
+pub use crate::error::{FStringErrorType, LexicalErrorType, ParseError, ParseErrorType};
 pub use crate::token::{Token, TokenKind};
 
 use crate::parser::Parser;
 
-use ruff_python_ast::{Expr, Mod, ModExpression, ModModule, PySourceType, Suite};
+pub use ruff_python_ast::{Expr, Mod, ModExpression, ModModule, PySourceType, Suite};
```
Comment (MichaReiser) [inline crates/ruff_python_parser/src/lib.rs:75]: This should not be necessary. You can use the `ruff_python_ast` directly

File: crates/ruff_python_parser/src/lib.rs [hunk 2]
```diff
@@ -234,7 +234,7 @@ pub fn parse_unchecked_source(source: &str, source_type: PySourceType) -> Parsed
 }
 
 /// Represents the parsed source code.
-#[derive(Debug, PartialEq, Clone)]
+#[derive(Debug, PartialEq, Clone, Default)]
```
Comment (MichaReiser) [inline crates/ruff_python_parser/src/lib.rs:237]: Can you explain me the use case for adding a `Default` implementation for `Parsed` and `Tokens` ?

File: crates/ruff_python_ast/src/nodes.rs [hunk 3]
```diff
@@ -27,8 +27,15 @@ pub enum Mod {
...
+impl Default for Mod {
+    #[inline]
+    fn default() -> Self {
+        Self::Module(ModModule::default())
+    }
+}
+
 /// See also [Module](https://docs.python.org/3/library/ast.html#ast.Module)
-#[derive(Clone, Debug, PartialEq)]
+#[derive(Clone, Debug, PartialEq, Default)]
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/nodes.rs:38]: Same here, it's unclear what the use case for `Default` is.

**Review by MichaReiser — CHANGES_REQUESTED** [review]
Hi. Nice that you're finding our parser useful. 

I'm open to making `LexicalErrorType` public but I rather not make the other changes.

**Comment by zao111222333** [comment]
> Hi. Nice that you're finding our parser useful.
> 
> I'm open to making `LexicalErrorType` public but I rather not make the other changes.

Hi, thank you for the suggestions. 
I need to store the tokens (`Tokens`/`Vec<Token>`) and errors (`Vec<ParseError>`), the initalization is required at beginning, so I add `Defualt` impl for some internal structs in previous commit.
And now I change to access the `Vec<Token>` and `Vec<ParseError>`, so that I can store the `Vec` and init with `Vec::new()`.
Here is my update

**Comment by MichaReiser** [comment]
> Hi, thank you for the suggestions.
> I need to store the tokens (Tokens/Vec<Token>) and errors (Vec<ParseError>), the initalization is required at beginning, so I add Defualt impl for some internal structs in previous commit.
> And now I change to access the Vec<Token> and Vec<ParseError>, so that I can store the Vec and init with Vec::new().
> Here is my update

Do you have a link to some code that I can look at? I would prefer to keep those types as private as possible.

**Comment by zao111222333** [comment]
> Do you have a link to some code that I can look at? I would prefer to keep those types as private as possible.

Sure, pls have a look at:
+ Store the `tokens` & `errors`, update them once the input code is changed
https://github.com/zao111222333/pyapp/blob/main/src/app.rs#L98-L99
https://github.com/zao111222333/pyapp/blob/main/src/app.rs#L163-L172
+ Incomplete detect
https://github.com/zao111222333/pyapp/blob/main/src/app.rs#L112-L151
+ Static code highlight
https://github.com/zao111222333/pyapp/blob/main/src/app.rs#L206-L364

**Comment by zao111222333** [comment]
btw, I have modified my commit, which I believe is not as rude as before. Pls see
https://github.com/astral-sh/ruff/pull/13509/commits/11ab0dbbdb1d901244e6ae941a35eaf86cd2a752

**Comment by MichaReiser** [comment]
Thanks for sharing your code. I don't think we need most of these changes.

I suggest that you:

* Change `MyHelper` to store a `Parsed` 
* You can call `parse_unchecked_source("")` to get a "default" parsed. This is slightly more expensive than calling `Parsed::default` (which doesn't exist) but shouldn't be that expensive because the lexer and parser will exit immediately (it's an empty string)

We then only need to make `LexicalErrorType` public which I think is a good change.

**Comment by MichaReiser** [comment]
And the repl looks very impressive!

**Review by MichaReiser — APPROVED** [review]

**Comment by zao111222333** [comment]
Thank you for your suggestion! I also believe that is a good way to get default parsed. And here is the changes as your suggestion.
Thanks again :)

**Comment by MichaReiser** [comment]
Thank you. This looks good

**Pull request merged**